src/main/java/frc/robot/subsystems/SwerveSubsystem.java

```diff
@@ -0,0 +1,191 @@
+import edu.wpi.first.math.kinematics.SwerveDriveOdometry;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.util.datalog.BooleanLogEntry;
+import edu.wpi.first.util.datalog.DataLog;
+import edu.wpi.first.util.datalog.DoubleLogEntry;
+import edu.wpi.first.wpilibj.DataLogManager;
+import edu.wpi.first.wpilibj.SPI;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import edu.wpi.first.wpilibj2.command.WaitUntilCommand;
+import frc.robot.Robot;
+import frc.robot.Constants.DriveConstants;
+import frc.robot.Constants.SwerveModuleConstants;
+import frc.robot.hardware.AbsoluteEncoders;
+import frc.robot.hardware.MotorController.MotorConfig;
+
+public class SwerveSubsystem extends SubsystemBase{
+    private final SwerveModule frontLeft = new SwerveModule(
+        MotorConfig.FrontLeftModuleDrive,
+        MotorConfig.FrontLeftModuleTurn,
+        AbsoluteEncoders.FrontLeftModule,
+        "FL");
+
+    private final SwerveModule frontRight = new SwerveModule(
+        MotorConfig.FrontRightModuleDrive,
+        MotorConfig.FrontRightModuleTurn,
+        AbsoluteEncoders.FrontRightModule,
+        "FR");
+
+    private final SwerveModule backLeft = new SwerveModule(
+        MotorConfig.BackLeftModuleDrive,
+        MotorConfig.BackLeftModuleTurn,
+        AbsoluteEncoders.BackLeftModule,
+        "BL");
+
+    private final SwerveModule backRight = new SwerveModule(
+        MotorConfig.BackRightModuleDrive,
+        MotorConfig.BackRightModuleTurn,
+        AbsoluteEncoders.BackRightModule,
+        "BR");
+
+    private AHRS gyro = new AHRS(SPI.Port.kMXP);
+    private SwerveDriveOdometry odometer = new SwerveDriveOdometry(DriveConstants.kDriveKinematics, new Rotation2d(0));
+
+    private DataLog datalog = DataLogManager.getLog();
+    private DoubleLogEntry translationXOutputLog = new DoubleLogEntry(datalog, "/swerve/txout"); //Logs x translation state output
+    private DoubleLogEntry translationYOutputLog = new DoubleLogEntry(datalog, "/swerve/tyout"); //Logs y translation state output
+    private DoubleLogEntry rotationOutputLog = new DoubleLogEntry(datalog, "/swerve/rotout"); //Logs rotation state output
+    private BooleanLogEntry controlOrientationLog = new BooleanLogEntry(datalog, "/swerve/orientation"); //Logs if robot is in FOD/ROD
+
+    public boolean controlOrientationIsFOD;
+
+    public SwerveSubsystem() {
+        new WaitUntilCommand(this::gyroReady)
+        .andThen(new InstantCommand(this::zeroHeading,this))
+        .schedule();
+        controlOrientationIsFOD = true;
+    }
+
+    private boolean gyroReady() {
+        return !gyro.isCalibrating();
+    }
+
+    public void zeroHeading() {
+        gyro.reset();
+    }
+
+    public double getHeading() {
+        return Math.IEEEremainder(gyro.getAngle(), 360); //Clamps angle output between -180 and 180 degrees
```

Comment:
I think it's correct, it's -180 to 180 in simulation and I can't get a number outside of that range using this [calculator](http://www.xcalculate.com/Calculator/30)

Reply:
You're right, though it might be worth adding a comment explaining what IEEEremainder does (or just link to a docs page like [this one](https://learn.microsoft.com/en-us/dotnet/api/system.math.ieeeremainder?view=net-7.0)) since it doesn't work the way one might expect a remainder function to.  

That link it technically for C# but the java docs pages I found don't seem to explain it as succinctly.